tests/kafkatest/tests/connect/connect_plugins_rest_test.py

```diff
@@ -0,0 +1,250 @@
+    }
+
+    FILE_SINK_CONFIGS = {
+        'file': {'required': True, 'type': 'STRING', 'importance': 'HIGH', 'group': None, 'order': -1, 'width': 'NONE'}
+    }
+
+    FILE_SOURCE_ALL_CONFIGS = merge_two_dicts(COMMON_CONNECTOR_CONFIGS, FILE_SOURCE_CONFIGS)
+    FILE_SINK_ALL_CONFIGS = merge_two_dicts(COMMON_CONNECTOR_CONFIGS, FILE_SINK_CONFIGS)
+
+    FILE_SOURCE_CONNECTOR = 'org.apache.kafka.connect.file.FileStreamSourceConnector'
+    FILE_SOURCE_CONNECTOR_SHORT = 'FileStreamSourceConnector'
+    FILE_SOURCE_CONNECTOR_ALIAS = 'FileStreamSource'
+
+    FILE_SINK_CONNECTOR = 'org.apache.kafka.connect.file.FileStreamSinkConnector'
+    FILE_SINK_CONNECTOR_SHORT = 'FileStreamSinkConnector'
+    FILE_SINK_CONNECTOR_ALIAS = 'FileStreamSink'
+
+    def __init__(self, test_context):
+        super(ConnectPluginsRestApiTest, self).__init__(test_context, num_zk=1, num_brokers=1)
+        self.cc = ConnectDistributedService(test_context, 2, self.kafka)
+
+    def test_rest_api(self):
+        self.cc.set_configs(lambda node: self.render("connect-distributed.properties", node=node))
+
+        self.cc.start()
+
+        expected = {self.FILE_SOURCE_CONNECTOR, self.FILE_SINK_CONNECTOR}
+        connector_plugins = self.cc.list_connector_plugins()
+        connector_plugins_set = set([connector_plugin['class'] for connector_plugin in connector_plugins])
+        self.validate_plugins(expected, connector_plugins_set)
+        self.validate_config_def(self.FILE_SOURCE_CONNECTOR, self.COMMON_CONNECTOR_CONFIGS, {})
+        self.validate_config_def(self.FILE_SOURCE_CONNECTOR_SHORT, self.COMMON_CONNECTOR_CONFIGS, {})
+        self.validate_config_def(self.FILE_SOURCE_CONNECTOR_ALIAS, self.COMMON_CONNECTOR_CONFIGS, {})
+
+        self.validate_config_def(self.FILE_SINK_CONNECTOR, self.COMMON_CONNECTOR_CONFIGS, {})
+        self.validate_config_def(self.FILE_SINK_CONNECTOR_SHORT, self.COMMON_CONNECTOR_CONFIGS, {})
+        self.validate_config_def(self.FILE_SINK_CONNECTOR_ALIAS, self.COMMON_CONNECTOR_CONFIGS, {})
+
+        # Test the case that we have invalid values in connector common configs
+        props = {
+            'name': 'test',
+            'connector.class': self.FILE_SOURCE_CONNECTOR,
+            'tasks.max': 0
+        }
+
+        expected_config_values = {
+            'name': {'name': 'name', 'value': 'test', 'errors': [], 'recommended_values': [], 'visible': True},
+            'connector.class': {'name': 'connector.class', 'value': self.FILE_SOURCE_CONNECTOR, 'errors': [], 'recommended_values': [], 'visible': True},
+            'tasks.max': {'name': 'tasks.max', 'value': 0, 'errors': ['Invalid value 0 for configuration tasks.max: Value must be at least 1'], 'recommended_values': [], 'visible': True},
+            'topics': {'name': 'topics', 'value': [], 'errors': [], 'recommended_values': [], 'visible': True}
+        }
+
+        configs = self.validate_config_def(self.FILE_SOURCE_CONNECTOR, self.FILE_SOURCE_ALL_CONFIGS, props)
```

Comment:
@Ishiihara Right, testing that it actually works with the plugins is good. I worry about testing a ton of details like matching every single field in the output -- that's better tested and iterated on via unit tests. If we evolve the format, e.g. adding new fields, it's more painful to validate that via system tests.

I'm not necessarily worried about adding one test (especially when it doesn't have an parameterized versions). I ask about this because system tests aren't like unit tests, so they should often encompass more functionality. Some of the stuff you're checking here, like error counts, can be checked in unit tests of AbstractHerder (where this count is computed).

You're right that checking actual instantiation after validating (or failing validation of) the config would also be good. This would actually fit quite nicely with the existing REST test, which also just exercises basic listing, instantiation, configs, and deletion. That test has a similar issue in that it is checking exact equivalence of expected output from the API; these checks only worry me a bit more because there's so much more to them. Might make sense to live with those more extensive checks for now, at the cost of some fragility, but just refactor some of this into the existing REST test.